update for the implementation is changed (#1)
Fullstop000/rfcs: text/2019-11-13-follower-replication.md

## Summary

> This RFC introduces a new mechanism into Raft protocol which allows a follower to send raft logs and snapshots to other followers. The target of this feature is to reduce network transmission cost between different data centers in Log Replication.

Reviewer: ```suggestion
This RFC introduces a new mechanism for Raft protocol, which allows a follower to replicate raft logs and snapshots to other followers. The target of this feature is to reduce network transmission costs between the followers in different data centers in Log Replication.
```

## Motivation

In the origin Raft protocol, a follower can only receive raft logs and snapshots from its leader, which could be insufficient in some situations. For example, when a Raft group is distributed in different data centers, log replication between a new node and the leader could be expensive if they are in different data centers. In this case, internal follower-to-follower transfer in one data center can be far more efficient than the traditionally stipulated leader-to-follower transfer. And in a cluster with massive nodes, the leader can have lower system load due to the less operations in network transmission.

## Detailed design

There are four key concepts of design:

- Every peer in a Raft group is associated with a `group_id`.
- Follower-to-follower data transfer is only allowed for peers in same group.

> - Peers report their `group_id` to their leader in Raft messages.

Reviewer: ```suggestion
- Peers report their `group_id` to their leader by Raft messages.
```

As the `Groups` is configured, the leader is able to choose a group member as a *delegate* to send entries to one or the rest group members in Log Replication, which is called Follower Replication. To implement Follower Replication, we basically need three main steps:

1. The leader picks a group member as a delegate of the group.

> 2. The leader tells the delegate the name list of other peers in the same group with the delegate.

Reviewer: ```suggestion
2. When processing Log Replication to a delegate, the leader must specify peers whom the delegate should replicate logs to. And the leader never sends raft logs to the specified peers.
```

> 3. the delegate replicates new appended entries to those peers.

Reviewer: ```suggestion
3. The delegate replicates entries to specified peers based on its own progress.
```

    1. The progress has the maximal `match_index`.

3. If no delegate is picked, the leader does Log Replication itself. Especially, if a group contains the leader it self, no delegate will be set. And the feature of enabling choosing delegate of such 'leader group' can be controlled by upper layer.

If the delegate of a group is determined, it’ll be cached in memory for later usage until Groups configuration is changed or the delegate is evicted by rejection message. And the flow control should be valid for any cached delegate.

#### How delegate does replication

We can add a new field into `Message`: `BcastTargets`, which is a list of peer IDs that the message receiver (the delegate) needs to replicate entries to. When a peer become delegate first time, it doesn't know followers' progress, so it appends its last log entry to other peers. After it receives `MsgAppendResponse` from those peers, it can update their progresses.

You can just treat a commission as the metadata of a MsgAppend or MsgSnapshot since the `last_index` and `log_term` are set by the leader according to its progress set.

> Peers can know `MsgAppend` comes from a delegate or its leader, because an another field `delegate` is added in `Message`. If a peer receives a `MsgAppend` from a delegate, it needs to response both its leader and the delegate, so that both its leader and delegate can update progress for it.

Reviewer: ```suggestion
Peers can know whether a `MsgAppend` comes from a delegate or the leader according to `delegate` value in the message. If a peer receives a `MsgAppend` from a delegate, it needs to respond to both the leader and the delegate to sync the `Progress` on them.
```

## Drawbacks

1. Committing speed could be slow if `quorum` must involve peers in different data center.

> 2. When handle a Raft message, a hashmap query on `map[peer_id]group_id` is introduced.

Reviewer: ```suggestion
2. When handling a Raft message, a hashmap query to get a peer's group and delegate  is introduced.
```

> 

Reviewer: ```suggestion
3. When a delegate is dismissed due to networking partition, the leader recovers processing Log Replication to peers in the delegate's group. But the information in `Inflights` on the delegate won't be extended by the leader.
Therefore, at that time, the leader starts to send messages with a plain `Inflights`, which breaks the origin flow control. And vice versa in the situation where the leader picks a delegate.

```

## Alternatives

> 

Reviewer: ```suggestion
In the current design, the leader not only let the delegate replicate logs to other peers but also 'delegate's the whole flow control to the delegate. We can benefit from it because of the lighter cross-data-center network transmission between the leader and peers but suffer from some issues described in **Drawbacks** section.
There is an alternative design that the leader doesn't 'delegate' the flow control to the delegate. When the leader sends `BcastTargets` to the delegate, it adds the `last_index` of each member in `BcastTargets` to indicate the peer's progress from the leader's view. In this way, the messaging flow is always controlled by the leader and the `Inflights` in-consistency can be avoided. 
```
